src/main/webapp/one-off/databaseInit.java

```diff
@@ -40,104 +39,172 @@
+import com.google.appengine.api.search.Results;
+import com.google.appengine.api.search.ScoredDocument;
+import com.google.appengine.api.search.SearchException;
+import com.google.appengine.api.search.SearchServiceFactory;
+import com.google.appengine.api.search.StatusCode;
 import com.google.gson.Gson;
+import java.io.*;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.*;
+import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
-import javax.servlet.ServletException;
-import java.util.*;
-import java.io.*;
-import java.nio.charset.StandardCharsets;
-import java.nio.file.Paths;
-import java.nio.file.Files;
+import org.apache.commons.lang3.StringUtils;
 
-
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/test")
 public class TestUploadServlet extends HttpServlet {
+  private String RECIPES_DIRECTORY = "RECIPES_DIRECTORY";
+  private String RECIPE_FILE_EXTENSION = ".txt";
+
+  /**
+   * Receives a recipe's properties and creates an entity, for the database and a document for the index
+   */
+  public void upload(
+    String title,
+    String imgURL,
+    ArrayList<String> ingredients,
+    ArrayList<String> stepList
+  )
+    throws Exception, InterruptedException, IOException {
+    IndexSpec indexSpec = IndexSpec
+      .newBuilder()
+      .setName("recipesIndex")
+      .build();
+    Index index = SearchServiceFactory.getSearchService().getIndex(indexSpec);
 
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     /**
-      * upload takes the several properties of a recipes, uses them in creating a new recipe entity and upload thi entity to the database
-     **/
-    public void upload(String  title, String imgURL, ArrayList<String> ingredients, ArrayList<String> stepList) throws Exception, IOException {
-        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    
-        Entity recipeEntity = new Entity("Recipe");
-
-        recipeEntity.setProperty("title", title);
-        recipeEntity.setProperty("index_title", title.toLowerCase());
-        recipeEntity.setProperty("imgURL", imgURL);
-        recipeEntity.setProperty("ingredients", ingredients);
-        recipeEntity.setProperty("stepList", stepList);
-        recipeEntity.setProperty("author", "Piece of Cake");
-        recipeEntity.setProperty("difficulty", "N/A");
-        recipeEntity.setProperty("prep_time", "N/A");
-        recipeEntity.setProperty("cook_time", "N/A");
-
-        datastore.put(recipeEntity);
-    }
+     * We alocate an id here because we need it inside the index documents
+     * This will help us to search for the recipes in an efficient way
+     */
```

Comment:
These kind of comments are only used for documentation, use // comments here, even if they span multiple lines.

Also the comment is explaining the functionality but not the motivation of the id. It should say "The recipes are going to be accessed by the ID, henceforth we're reserving the ID here, setting it in the entity and saving it in the index. When we search for a recipe for any of its indexed attributes, the ID is going to be available for fetching the rest of the data from the datastore."

Also the ID here is not helping to "search the recipes in a efficient way", but to actually be able to load the data from the datastore for that recipe.

Try to think from the point of view of someone that's reading this piece of code for the first time, it should be possible to understand the WHY and HOW from the comment.

Reply:
Ok, this is really helpful, thank you!